clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -547,16 +580,29 @@ public void onFailure(RuntimeException e) {
      *
      * @param node The node to send the ListOffsetRequest to.
      * @param timestampsToSearch The mapping from partitions to the target timestamps.
+     * @param requireTimestamp  True if we require a timestamp in the response.
      * @return A response which can be polled to obtain the corresponding timestamps and offsets.
      */
     private RequestFuture<Map<TopicPartition, OffsetAndTimestamp>> sendListOffsetRequest(final Node node,
-                                                                                         final Map<TopicPartition, Long> timestampsToSearch) {
-        ListOffsetRequest request = new ListOffsetRequest(timestampsToSearch, ListOffsetRequest.CONSUMER_REPLICA_ID);
-        log.trace("Sending ListOffsetRequest {} to broker {}", request, node);
-        return client.send(node, ApiKeys.LIST_OFFSETS, request)
+                                                                                         final Map<TopicPartition, Long> timestampsToSearch,
+                                                                                         boolean requireTimestamp) {
+        ListOffsetRequest.Builder builder = new ListOffsetRequest.Builder().
+                setTargetTimes(timestampsToSearch).
+                setReplicaId(ListOffsetRequest.CONSUMER_REPLICA_ID);
+
+        // If we need a timestamp in the response, the minimum RPC version we can send is v1.
+        // Otherwise, v0 is OK.
+        builder.setMinVersion(requireTimestamp ? (short) 1 : (short) 0);
```

Comment:
I'm thinking of the case where the broker doesn't support v1 of ListOffsets. For this case, I think we currently raise `ObsoleteBrokerException`. I am questioning whether it would be more consistent to return a null entry in the result of `offsetsForTimes`. Currently it is possible for the broker to support the new api version, but not the message format version which is needed to answer the query. In this case, we return a null entry.